lib/ssh/src/ssh_file_with_agent.erl

```diff
@@ -0,0 +1,62 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2019. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+
+%%% Description: SSH file handling with SSH Agent support
+
+-module(ssh_file_with_agent).
+
+-behaviour(ssh_client_key_api).
+
+-include("ssh.hrl").
+-include("ssh_agent.hrl").
+
+-export([add_host_key/3, is_host_key/4, user_key/2]).
+
+add_host_key(Host, Key, Opts) ->
+    ssh_file:add_host_key(Host, Key, Opts).
+
+is_host_key(Key, PeerName, Algorithm, Opts) ->
+    ssh_file:is_host_key(Key, PeerName, Algorithm, Opts).
+
+user_key(Algorithm, _Opts) ->
+    Request = #ssh_agent_identities_request{},
+    Response = ssh_agent:send(Request),
+
+    #ssh_agent_identities_response{keys = Keys} = Response,
+
+    AlgorithmStr = atom_to_list(Algorithm),
+    MatchingKeys = lists:filter(fun(Key) -> has_key_type(Key, AlgorithmStr) end, Keys),
+
+    % The "client_key_api" behaviour only allows returning a single user key,
+    % so we simply select the first one returned from the SSH agent here. This
+    % means that if a user adds multiple keys for the same algorithm, only the
+    % first one added will be used.
+    [#ssh_agent_key{blob = PubKeyBlob} | _OtherKeys] = MatchingKeys,
+
+    % TODO:
+    % Mark the returned key so we can identify agent keys as they do not
+    % contain the private key.
+    {ok, PubKeyBlob}.
```

Comment:
This does not strictly conform to the interface outlined by the `ssh_client_key_api` behavior. The function is expected to return a [`public_key:private_key()`](http://erlang.org/doc/man/ssh_client_key_api.html#Module:user_key-2) but we're returning a binary. My idea for handling this is to add a `SSHAgentKey` type to the list of acceptable `public_key:private_key()`s to make the kind of key more explicit and to facilitate matching in `ssh_auth`.

Reply:
Alternatives could be

- to introduce a completely new option to `ssh:connect`
- to tag the value provided for the `key_cb` option
- to create an alternative to the `ssh_client_key_api` behaviour and `try/catch` to either use the new one or fallback to the current `ssh_client_key_api` behaviour based on what callbacks are defined/callable

Reply:
I'd be escpecially happy about feedback/ideas on how to best tackle this ✌️

Reply:
I have no idea right now, just some thoughts.  The existing api prescribes records, one per key type. But the second thing ssh_auth does is to encode the key into a blob like the one that ssh_agent returns. As a side note, the correctness checking that ssh_auth does is due to an error report where a key was placed in a file with the wrong name.
For the moment I think that tagging the blob with something descriptive and general not tied to ssh_agent would be the best.  For example {ssh2_pubkey, binary()} or something better.